Server has state "STOPPED", but it's not

[kind/bug]

<!--

Welcome! - We kindly ask you to:

  1. Fill out the issue template below 

Thanks for understanding, and for contributing to the project!

-->

## What versions of software are you using?
- Linux (Fedora 33)
- From "master" branch: https://github.com/redhat-developer/intellij-rsp 
- From "master" branch: https://github.com/redhat-developer/intellij-rsp 

## What did you run exactly?

./gradlew runIde -> Start Community Server Connector by Red Hat -> download tomcat 10.0.8 -> start server -> stop server -> close IDE -> see error -> click "Terminate" -> IDE is frozen

## Actual behavior

![stop_server_intellij_rsp_error](https://user-images.githubusercontent.com/60704619/222120999-a9cbbf23-af1c-4638-a74a-b4c01899a23e.png)

## Expected behavior

No errors.

## Any logs, error output, etc?
`2023-03-01 12:07:49,111 [  27184]   WARN - ion.process.UnixProcessManager - Error killing the process 
java.lang.IllegalStateException: Failed to get PID from an instance of class com.redhat.devtools.intellij.rsp.util.RemoteServerProcess, OS: Linux
        at com.intellij.execution.process.UnixProcessManager.getProcessId(UnixProcessManager.java:66)
        at com.intellij.execution.process.UnixProcessManager.sendSignalToProcessTree(UnixProcessManager.java:116)
        at com.intellij.execution.process.UnixProcessManager.sendSigKillToProcessTree(UnixProcessManager.java:111)
        at com.intellij.execution.process.OSProcessUtil.killProcessTree(OSProcessUtil.java:47)
        at com.intellij.execution.process.OSProcessHandler.killProcessTreeSync(OSProcessHandler.java:249)
        at com.intellij.execution.process.OSProcessHandler.lambda$killProcessTree$1(OSProcessHandler.java:243)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
        at java.base/java.util.Objects.requireNonNull(Objects.java:221)
        at com.intellij.execution.process.UnixProcessManager.getProcessId(UnixProcessManager.java:62)
        ... 14 more
2023-03-01 12:07:49,112 [  27185]   WARN - ution.process.OSProcessHandler - Process has been already terminated: Server apache-tomcat-10.0.8 
`

![intellij-rsp-stop](https://user-images.githubusercontent.com/60704619/222121245-e8fa8e9a-93db-4f4b-ac62-81b1c5b78366.gif)

--- a/.github/workflows/IJ.yml
+++ b/.github/workflows/IJ.yml
@@ -5,25 +5,49 @@ name: Validate against IJ versions
 
 on:
   push:
-    branches: [ master ]
+    branches: [ main ]
   pull_request:
-    branches: [ master ]    
+    branches: [ main ]
 
 jobs:
   validate:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        IJ: [IC-2019.3, IC-2020.1, IC-2020.2, IC-2020.3, IC-2021.1, IC-2021.2, IC-2021.3]
+        IJ: [IC-2021.1, IC-2021.2, IC-2021.3, IC-2022.1, IC-2022.2, IC-2022.3]
 
     steps:
     - uses: actions/checkout@v2
-    - name: Set up JDK 11
-      uses: actions/setup-java@v1
+    - name: Set up JDK 17
+      uses: actions/setup-java@v3
       with:
-        java-version: 11
+        java-version: 17
+        distribution: 'temurin'
+        cache: 'gradle'
     - name: Grant execute permission for gradlew
       run: chmod +x gradlew
     - name: Build with Gradle
       run: ./gradlew build -PideaVersion=${{ matrix.IJ }}
 
+  verify:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout Code
+        uses: actions/checkout@v2
+      - name: Set up JDK 17
+        uses: actions/setup-java@v3
+        with:
+          java-version: 17
+          distribution: 'temurin'
+          cache: 'gradle'
+      - name: Grant execute permission for gradlew
+        run: chmod +x gradlew
+      - name: Build with Gradle
+        run: ./gradlew runPluginVerifier -PideaVersion=IC-2022.3
+      - name: Upload report
+        uses: actions/upload-artifact@v2
+        if: always()
+        with:
+          name: verifier-report
+          path: build/reports/pluginVerifier

--- a/src/main/java/com/redhat/devtools/intellij/rsp/model/impl/SingleRspModel.java
+++ b/src/main/java/com/redhat/devtools/intellij/rsp/model/impl/SingleRspModel.java
@@ -132,8 +132,9 @@ public class SingleRspModel {
         return id;
     }
     public PtyProcess addServerProcess(ServerProcess serverProcess) {
-        RemoteServerProcess sp = new RemoteServerProcess();
-        processes.put(internalIdForProcess(serverProcess), sp);
+        String id = internalIdForProcess(serverProcess);
+        RemoteServerProcess sp = new RemoteServerProcess(id);
+        processes.put(id, sp);
         return sp;
     }
 
@@ -141,7 +142,7 @@ public class SingleRspModel {
         String id = internalIdForProcess(serverProcess);
         RemoteServerProcess sp = processes.get(id);
         if( sp != null ) {
-            sp.terminate();
+            sp.setTerminating();
             processes.remove(id);
         }
     }

--- a/src/main/java/com/redhat/devtools/intellij/rsp/util/RemoteServerProcess.java
+++ b/src/main/java/com/redhat/devtools/intellij/rsp/util/RemoteServerProcess.java
@@ -10,6 +10,7 @@
  ******************************************************************************/
 package com.redhat.devtools.intellij.rsp.util;
 
+import com.intellij.execution.process.SelfKiller;
 import com.pty4j.PtyProcess;
 import com.pty4j.WinSize;
 import java.util.Objects;
@@ -25,31 +26,35 @@ import java.util.concurrent.LinkedBlockingQueue;
  * Used to wrap output from a remote server process into something
  * usable by the terminal view.
  */
-public class RemoteServerProcess extends PtyProcess {
+public class RemoteServerProcess extends PtyProcess implements SelfKiller {
     private boolean terminated = false;
+    private boolean terminating = false;
     private OutputStream serverSysIn;
     private PipedOutputStream serverSysOutInternal;
     private PipedOutputStream serverSysErrInternal;
     private PipedInputStream serverSysOut;
     private PipedInputStream serverSysErr;
     private RSPThread writer;
-    public RemoteServerProcess() {
-        serverSysIn = new OutputStream() { @Override public void write(int b) { } };
-        serverSysOut =new PipedInputStream();
-        serverSysErr =new PipedInputStream();
+    private String name;
+
+    public RemoteServerProcess(String name) {
+        serverSysIn = new OutputStream() {
+            @Override
+            public void write(int b) {
+            }
+        };
+        serverSysOut = new PipedInputStream();
+        serverSysErr = new PipedInputStream();
         serverSysOutInternal = new PipedOutputStream();
         serverSysErrInternal = new PipedOutputStream();
         try {
             serverSysOut.connect(serverSysOutInternal);
             serverSysErr.connect(serverSysErrInternal);
-        } catch(IOException ioe) {
+        } catch (IOException ioe) {
             // TODO ?
+            ioe.printStackTrace();
         }
-    }
-
-    @Override
-    public boolean isRunning() {
-        return !isTerminated();
+        this.name = name;
     }
 
     @Override
@@ -69,15 +74,16 @@ public class RemoteServerProcess extends PtyProcess {
 
     private class RSPThread extends Thread {
         private BlockingQueue<ServerProcessOutput> queue;
+
         public RSPThread() {
             super("Server Process Output Processor");
             queue = new LinkedBlockingQueue<ServerProcessOutput>();
         }
+
         public void run() {
             try {
                 while (true) {
-                    if( queue.size() == 0 && isTerminated()) {
-                        cleanup();
+                    if (queue.size() == 0 && isTerminating()) {
                         return;
                     }
                     ServerProcessOutput serverProcessOutput = queue.poll(1, TimeUnit.SECONDS);
@@ -87,16 +93,20 @@ public class RemoteServerProcess extends PtyProcess {
                 }
             } catch (InterruptedException e) {
                 Thread.currentThread().interrupt();
+            } finally {
+                cleanup();
+                setTerminated();
             }
         }
+
         public void addElement(ServerProcessOutput out) {
-            if( out != null )
+            if (out != null)
                 queue.add(out);
         }
     }
 
     public void handleEvent(ServerProcessOutput output) {
-        if( writer == null ) {
+        if (writer == null) {
             writer = new RSPThread();
             writer.start();
         }
@@ -109,9 +119,9 @@ public class RemoteServerProcess extends PtyProcess {
                 serverSysOutInternal.write(output.getText().getBytes());
             }
             if (output.getStreamType() == ServerManagementAPIConstants.STREAM_TYPE_SYSERR) {
-                serverSysOutInternal.write(output.getText().getBytes());
+                serverSysErrInternal.write(output.getText().getBytes());
             }
-        } catch(IOException ioe) {
+        } catch (IOException ioe) {
             // TODO
         }
     }
@@ -134,70 +144,92 @@ public class RemoteServerProcess extends PtyProcess {
     @Override
     public int waitFor() throws InterruptedException {
         // TODO fix this
-        while( !isTerminated()) {
+        while (!isTerminated()) {
             try {
                 Thread.sleep(1000);
-            } catch(InterruptedException ie) {
+            } catch (InterruptedException ie) {
                 throw ie;
             }
         }
         return 0;
     }
 
+
+    @Override
+    public boolean isRunning() {
+        return !isTerminated();
+    }
+
+    @Override
+    public boolean isAlive() {
+        return !isTerminated();
+    }
+
     @Override
     public int exitValue() {
-        if( isTerminated() )
+        if (isTerminated())
             return 0;
         throw new IllegalThreadStateException("Server not terminated yet");
     }
 
     @Override
     public void destroy() {
-
+        setTerminating();
+        cleanup();
+        setTerminated();
     }
 
-    public synchronized void terminate() {
-        terminated = true;
-        // necessary to delay "terminate" to allow time for output to show
-        // if no output shows when we mark terminated, the terminal closes quickly
-        new Thread("Delay Process Termination 2s") {
-            public void run() {
-                try {
-                    Thread.sleep(2000);
-                } catch(InterruptedException ie) {}
-                setTerminated();
-            }
-        }.start();
+    public synchronized void setTerminating() {
+        this.terminating = true;
+        if( this.writer == null ) {
+            cleanup();
+            setTerminated();
+        }
     }
 
     private synchronized void setTerminated() {
-        terminated = true;
+        this.terminated = true;
     }
+
     private void cleanup() {
         try {
             serverSysIn.close();
-        } catch(IOException ioe) {
+        } catch (IOException ioe) {
         }
         try {
             serverSysOut.close();
-        } catch(IOException ioe) {
+        } catch (IOException ioe) {
         }
         try {
             serverSysErr.close();
-        } catch(IOException ioe) {
+        } catch (IOException ioe) {
+        }
+        try {
+            serverSysOutInternal.flush();
+        } catch( IOException ioe) {
         }
         try {
             serverSysOutInternal.close();
-        } catch(IOException ioe) {
+        } catch (IOException ioe) {
+        }
+        try {
+            serverSysErrInternal.flush();
+        } catch( IOException ioe) {
         }
         try {
             serverSysErrInternal.close();
-        } catch(IOException ioe) {
+        } catch (IOException ioe) {
         }
-        writer.interrupt();
+        if( writer != null ) {
+            writer.interrupt();
+        }
     }
 
     private synchronized boolean isTerminated() {
-        return terminated;
+        return this.terminated;
     }
+    private synchronized boolean isTerminating() {
+        return this.terminating;
+    }
+
 }

